implement service-level extra_config configuration and validation

Strangerxxx opened the pull request.

Strangerxxx committed:
implement service-level extra_config configuration and validation

Strangerxxx committed:
implement proper version detect

tests are not compatible anymore

Strangerxxx committed:
update helm charts

Strangerxxx committed:
fix tests

dronov commented:
I got this message after installing krakend-operator chart (after installing new krakend-operator-crds):
```
k describe pod krakend-operator-controller-manager-6fc48df448-2277f
Name:             krakend-operator-controller-manager-6fc48df448-2277f
Namespace:        default
Priority:         0
Service Account:  krakend-operator-controller-manager
Node:             server4/192.168.1.105
Start Time:       Thu, 01 Aug 2024 16:59:13 +0200
Labels:           app.kubernetes.io/instance=krakend-operator
                  app.kubernetes.io/name=krakend-operator
                  control-plane=controller-manager
                  pod-template-hash=6fc48df448
Annotations:      kubectl.kubernetes.io/default-container: manager
Status:           Pending
SeccompProfile:   RuntimeDefault
IP:
IPs:              <none>
Controlled By:    ReplicaSet/krakend-operator-controller-manager-6fc48df448
Containers:
  manager:
    Container ID:
    Image:         europe-north1-docker.pkg.dev/nais-io/nais/images/krakend-operator:latest
    Image ID:
    Port:          9443/TCP
    Host Port:     0/TCP
    Command:
      /manager
    Args:
      --health-probe-bind-address=:8081
      --metrics-bind-address=127.0.0.1:8080
      --leader-elect
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Requests:
      cpu:      100m
      memory:   256Mi
    Liveness:   http-get http://:8081/healthz delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:  http-get http://:8081/readyz delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment:
      KRAKEND_CHART_PATH:         /var/config/krakendinstaller-1.0.0.tgz
      NETPOL_ENABLED:             true
      DEBUG:                      false
      KUBERNETES_CLUSTER_DOMAIN:  cluster.local
    Mounts:
      /tmp/k8s-webhook-server/serving-certs from cert (ro)
      /var/config from config (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-s972j (ro)
  kube-rbac-proxy:
    Container ID:
    Image:         gcr.io/kubebuilder/kube-rbac-proxy:v0.14.1
    Image ID:
    Port:          8443/TCP
    Host Port:     0/TCP
    Args:
      --secure-listen-address=0.0.0.0:8443
      --upstream=http://127.0.0.1:8080/
      --logtostderr=true
      --v=0
    State:          Waiting
      Reason:       ContainerCreating
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     500m
      memory:  128Mi
    Requests:
      cpu:     5m
      memory:  64Mi
    Environment:
      KUBERNETES_CLUSTER_DOMAIN:  cluster.local
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-s972j (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  cert:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  krakend-operator-webhook-server-cert
    Optional:    false
  config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      config
    Optional:  false
  kube-api-access-s972j:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason       Age               From               Message
  ----     ------       ----              ----               -------
  Normal   Scheduled    12s               default-scheduler  Successfully assigned default/krakend-operator-controller-manager-6fc48df448-2277f to server4
  Warning  FailedMount  5s (x5 over 12s)  kubelet            MountVolume.SetUp failed for volume "config" : configmap "config" not found
 ```
 What have I missed?

Strangerxxx committed:
fix configmap reference name

Strangerxxx commented:
since helm chart was regenerated with helmify out of manifests, build from kustomization directory config/default/, which does not contain missing ConfigMap manifest, the reference to the ConfigMap was overwritten

I just bring it back manually, but we must consider to reimplement it somehow to be able to regenerate chart loosing no references.

Strangerxxx committed:
webhook fixes

- properly initialize krakends webhook
- enable modify certDir for debug
- remove duplicated logs
- implement getversion job cleanup
- fix regex usage for version detection on krakends webhook

dronov committed:
update krakend-operator image

Strangerxxx committed:
enable batch/jobs access to manager

Strangerxxx committed:
revert unnecessary changes

dronov commented:
👍
